### DISCUSSION TITLE
Panel navigation

### DISCUSSION THREAD

**Discussion by martin-braun** [vscode-neovim/vscode-neovim#1252] [General]
Switching tabs and moving around in panels, like the terminal panels. Maybe it makes sense to provide reasonable defaults?

If so, what could those be?

<img width="1740" alt="Screenshot 2023-07-02 at 14 49 19" src="https://github.com/vscode-neovim/vscode-neovim/assets/24259245/88d98063-1246-4b89-b17c-79f29e82bdaf">

- Opening the terminal and focusing it
- Jumping between tabs
- Open new terminal panel
- Jumping between terminals
- Swaping order of terminals
- Resizing terminal panels
- Scrolling in terminal
- Zoom the terminal

For me this works (`ctrl+t` is my `tmux` prefix):

```json
    {
        "key": "ctrl+t z",
        "command": "workbench.action.toggleMaximizedPanel"
    },
    {
        "key": "ctrl+t shift+5",
        "command": "workbench.action.terminal.split",
        "when": "terminalFocus && terminalProcessSupported || terminalFocus && terminalWebExtensionContributedProfile"
    },
    {
        "key": "ctrl+t ctrl+left",
        "command": "workbench.action.terminal.resizePaneLeft",
        "when": "terminalFocus && terminalHasBeenCreated || terminalFocus && terminalProcessSupported"
    },
    {
        "key": "ctrl+t ctrl+right",
        "command": "workbench.action.terminal.resizePaneRight",
        "when": "terminalFocus && terminalHasBeenCreated || terminalFocus && terminalProcessSupported"
    },
    {
        "key": "ctrl+t ctrl+up",
        "command": "workbench.action.terminal.resizePaneUp",
        "when": "terminalFocus && terminalHasBeenCreated || terminalFocus && terminalProcessSupported"
    },
    {
        "key": "ctrl+t ctrl+down",
        "command": "workbench.action.terminal.resizePaneDown",
        "when": "terminalFocus && terminalHasBeenCreated || terminalFocus && terminalProcessSupported"
    },
    {
        "key": "ctrl+t alt+5",
        "command": "workbench.action.terminal.scrollUpPage",
        "when": "terminalFocus && terminalHasBeenCreated && !terminalAltBufferActive || terminalFocus && terminalProcessSupported && !terminalAltBufferActive"
    },
    {
        "key": "ctrl+t f5",
        "command": "workbench.action.terminal.scrollUpPage",
        "when": "terminalFocus && terminalHasBeenCreated && !terminalAltBufferActive || terminalFocus && terminalProcessSupported && !terminalAltBufferActive"
    },
    {
        "key": "ctrl+t f6",
        "command": "workbench.action.terminal.scrollDownPage",
        "when": "terminalFocus && terminalHasBeenCreated && !terminalAltBufferActive || terminalFocus && terminalProcessSupported && !terminalAltBufferActive"
    },
    {
        "key": "ctrl+t right",
        "command": "workbench.action.terminal.focusNextPane",
        "when": "terminalFocus && terminalHasBeenCreated || terminalFocus && terminalProcessSupported"
    },
    {
        "key": "ctrl+t left",
        "command": "workbench.action.terminal.focusPreviousPane",
        "when": "terminalFocus && terminalHasBeenCreated || terminalFocus && terminalProcessSupported"
    },
    {
        "key": "ctrl+t up",
        "command": "workbench.action.focusPreviousGroup",
        "when": "terminalFocus && terminalHasBeenCreated || terminalFocus && terminalProcessSupported"
    },
    {
        "key": "ctrl+t shift+2",
        "command": "terminal.focus"
    },
    {
        "key": "ctrl+t down",
        "command": "workbench.action.terminal.focus"
    }

```

(Swapping is not part of my mappings right now. Also can't replicate real `tmux`-alike scrolling, kinda sucks.)

If there are no reasonable defaults, we should drop those for good. Thoughts?

**Comment by theol0403** [comment]
Yeah, this is tricky. The problem is that many people use different tmux prefixes, and that terminal navigation is somewhat out of the scope of nvim. I've seen C-b, C-t, C-space, etc. Additionally, the default tmux nav bindings are not vim-like, so there isn't really a "standard". 

What does terminal nvim do? Does it just use C-w + the whole suite?

**Reply by martin-braun** [reply]
> What does terminal nvim do? Does it just use C-w + the whole suite?

That's what I try to figure out with this discussion. I don't use terminal buffers, I use `tmux`. ;)

(sooner or later someone might step in here to answer my request, because that person misses that behavior, let's wait for that, the reason why this is a discussion and not an issue)

**Reply by theol0403** [reply]
Looks like this is the default: https://neovim.io/doc/user/nvim_terminal_emulator.html
- by default, `<C-\><C-N>` enters "terminal normal mode", where C-w works
- common to bind escape to "normal mode"
- common to bind `<A-h>`/`<A-j>` to pane navigation 

Looks like the best we can realistically do is 
* add `<C-\><C-N>` which returns to the last pane  
* add `<A-h>`

For what it's worth, currently the best way to leave the terminal is <code>ctrl+\`</code>

**Reply by martin-braun** [reply]
@theol0403 Thanks for your findings, been testing it on macOS with German keyboard, seems unusable. Can't press `<C-\>`, because `\` requires to press `shift+opt+7`, but `ctrl+shift+opt+7` will not be accepted. We cannot replicate the behavior from the Terminal at all, there is no terminal mode.

I was also not able to exit terminal mode. Once I'm in terminal mode `C-w` won't work either. I've also seen some people remapping things. We could probably map pane navigation to `<A-..>` to have sane defaults.

Returning to the last "pane" on `<C-\><C-N>` is unfeasible. The map will not work due to varying layouts. Don't know what <code>ctrl+\`</code> comes from either. I vote for simply enabling pane navigation via `<A-..>`, leaving the terminal will just work with `<C-w>..` and then we are calling it a day.

**Reply by theol0403** [reply]
Huh, it's weird that the default neovim binding is so inaccessible. Even if it were, it's not a nice sequence to press. 

Yeah, A-h navigation is probably best. Maybe we should have it behind a simple config flag however. 

Just a random tidbit, you don't need to say if you already know, but `workbench.action.focusActiveEditorGroup` is a good way of returning to the editor from a panel.

**Reply by martin-braun** [reply]
Okay, we need to elaborate: There are panel tabs that can be navigated with `workbench.action.next/previousPanelView` (Problems / Output / Debug Console / Terminal) and there are terminal panes to jump through via `workbench.action.terminal.focusNext/PreviousPane`. On top of that we have pane resize commands `workbench.action.terminal.resizePaneUp/Down/Left/Right` and a pane split command `workbench.action.terminal.split` which splits in the logical direction depending on the position of the panel (left/bottom/right).

We should find bindings for all of that, but sadly, we face an issue once again for Mac users. `alt+l` is used to type `@` on German keyboard layouts, so we definitely cannot map `alt`. 

- Which modifier could we use? The only one that was left was `ctrl` (which will cause other conflicts, see below) or we find a reasonable prefix like `ctrl+w` for window navigation, but for pane(l) navigation. This is a tough one, but I lean towards a prefix instead of a modifier, because it will allow us to implement the resizing via `+/-/</>` (see below)
- <s>Which buffer navigations are defaults? I remapped my ones, so I kindly ask. We should use the buffer navigation bindings for `next/previousPanelView` to make it coherent, it shouldn't collide</s> (we can't use `gt`/`gT` or any tab commands, the user should map their own(?))
- `focusNext/PreviousPanel` will be `hjkl` with the chosen modifier or prefix, but it can't be `alt` as mentioned above (I was also thinking about `alt+n/p`, but `alt+n` prints `~` on my keyboard, so `alt` will definitely be no candidate)
- It can also not be `<C-h/j/k/l>`, because `<C-l>` re-renders the terminal by default (system wide)
- `resizePaneUp/Down/left/Right` could have `+/-/</>` with a prefix (if we prefer to choose one over a modifier), but will face the same localization issues as in #1257
- `split` could be `s` and `v` with a prefix, but the direction of the split will always be determined by VSCode based on the position of the panel

The problem on choosing a prefix with `ctrl` is the potential collision with a `tmux` prefix that was changed and we know there are not many candidates to pick from due to vim bindings, so this feels like a dead end.

**Reply by theol0403** [reply]
I see. Yeah, I'm pretty stumped. I think supporting german layout adds another wrench to the complexity.

**Reply by martin-braun** [reply]
@theol0403 The question is not about supporting it. Introducing `<A-..>` will effectively break the ability to input special characters for some people, so that's not an option.

**Comment by theol0403** [comment]
Ref: https://github.com/vscode-neovim/vscode-neovim/issues/1076

**Comment by martin-braun** [comment]
@theol0403 How likely is `:terminal` support in VSCode? Let's assume interactive buffers work some day, could this be a thing in the future?

**Reply by theol0403** [reply]
It is definitely possible, the main issue is just developer resources/time to refactor buffer manager. 

However, I can't comment on the performance, well-roundedness of :term.
